sky/serve/autoscalers.py

```diff
@@ -318,12 +319,49 @@ def get_latest_version_with_min_replicas(
     def select_outdated_replicas_to_scale_down(
             self, replica_infos: Iterable['replica_managers.ReplicaInfo']
     ) -> List[int]:
+        """Select outdated replicas to scale down."""
+
+        if self.update_mode == serve_utils.UpdateMode.ROLLING:
+            latest_ready_replicas = []
+            old_replicas = []
+            for info in replica_infos:
+                if info.version == self.latest_version:
+                    if info.is_ready:
+                        latest_ready_replicas.append(info)
+                else:
+                    old_replicas.append(info)
+
+            num_latest_ready_replicas = len(latest_ready_replicas)
+
+            # We compare to target_num_replicas instead of min_replicas, to
+            # guarantee better service quality. Since mixing traffic across
+            # old and latest versions are allowed in rolling update, this will
+            # not affect the time it takes for the service to updated to the
+            # latest version.
+            if num_latest_ready_replicas >= self.target_num_replicas:
+                # Once the number of ready new replicas is greater than or equal
+                # to the target, we can scale down all old replicas.
+                return [info.replica_id for info in old_replicas]
+            # If rolling update is in progress, we scale down old replicas
+            # based on the number of ready new replicas.
+            num_old_replicas_to_keep = (self.target_num_replicas -
+                                        num_latest_ready_replicas)
+
+            # Remove old replicas (especially old launching replicas) and only
+            # keep the required number of replicas, as we want to let the new
+            # replicas to take over the provisioning old replicas faster.
+            # `_select_replicas_to_scale_down` will make sure we scale the
+            # replicas in initializing statuses first before scaling down the
+            # READY old replicas.
+            return self._select_replicas_to_scale_down(
```

Comment:
Should we filter out failed old replicas first? iiuc the current impl will keep failed old replicas.

Reply:
Good point! It seems the latest master is problemetic then, as we are also terminating those failed old replicas first